### PR TITLE
Kxi 19292 conversion of arrow to kdb for nulls

### PR DESCRIPTION
**Pull request by vgrechin-kx**
Conversion of arrow to kdb for nulls

https://kxl.atlassian.net/browse/KXI-19292

### CODE REVIEW

File: src/ArrayReader.cpp [hunk 1]
```diff
@@ -159,7 +159,15 @@ template<>
 void AppendArray<arrow::Type::INT16>(shared_ptr<arrow::Array> array_data, K k_array, size_t& index, TypeMappingOverride& type_overrides)
 {
   auto int16_array = static_pointer_cast<arrow::Int16Array>(array_data);
-  memcpy(kH(k_array), int16_array->raw_values(), int16_array->length() * sizeof(arrow::Int16Array::value_type));
+  if( type_overrides.null_mapping.have_int16 ){
```
Comment (nmcdonnell-kx) [inline src/ArrayReader.cpp:162]:
```suggestion
  if( type_overrides.null_mapping.have_int16 && int16_array->null_count() != 0 ){

```

File: src/ArrayReader.cpp [hunk 2]
```diff
@@ -173,7 +181,15 @@ template<>
 void AppendArray<arrow::Type::INT32>(shared_ptr<arrow::Array> array_data, K k_array, size_t& index, TypeMappingOverride& type_overrides)
 {
   auto int32_array = static_pointer_cast<arrow::Int32Array>(array_data);
-  memcpy(kI(k_array), int32_array->raw_values(), int32_array->length() * sizeof(arrow::Int32Array::value_type));
+  if( type_overrides.null_mapping.have_int32 ){
```
Comment (nmcdonnell-kx) [inline src/ArrayReader.cpp:184]:
```suggestion
  if( type_overrides.null_mapping.have_int32 && int32_array->null_count() != 0 ){

```

File: src/ArrayReader.cpp (hunk 1 above)
Comment (vgrechin-kx) [reply src/ArrayReader.cpp:162]: good point - implementing

File: src/ArrayReader.cpp (hunk 2 above)
Comment (vgrechin-kx) [reply src/ArrayReader.cpp:184]: will propagate this change to other types where possible

File: src/ArrayReader.cpp [hunk 3]
```diff
@@ -130,96 +130,194 @@ void AppendArray<arrow::Type::BOOL>(shared_ptr<arrow::Array> array_data, K k_arr
...
+    memcpy(kJ(k_array), uint64_array->raw_values(), uint64_array->length() * sizeof(arrow::UInt64Array::value_type));
+  }
 }
 
 template<>
 void AppendArray<arrow::Type::INT64>(shared_ptr<arrow::Array> array_data, K k_array, size_t& index, TypeMappingOverride& type_overrides)
 {
   auto int64_array = static_pointer_cast<arrow::Int64Array>(array_data);
-  memcpy(kJ(k_array), int64_array->raw_values(), int64_array->length() * sizeof(arrow::Int64Array::value_type));
+  if( type_overrides.null_mapping.have_int32 && int64_array->null_count() ){
```
Comment (nmcdonnell-kx) [inline src/ArrayReader.cpp:249]:
```suggestion
  if( type_overrides.null_mapping.have_int64 && int64_array->null_count() ){

```

File: src/ArrayReader.cpp [hunk 4]
```diff
@@ -345,23 +489,37 @@ void AppendArray<arrow::Type::DURATION>(shared_ptr<arrow::Array> array_data, K k
...
+      + ( !( type_overrides.null_mapping.have_duration && dur_array->IsNull( i ) ) * tc.ArrowToKdb( dur_array->Value( i ) ) );
+  }
 }
 
 template<>
 void AppendArray<arrow::Type::INTERVAL_MONTHS>(shared_ptr<arrow::Array> array_data, K k_array, size_t& index, TypeMappingOverride& type_overrides)
 {
   auto month_array = static_pointer_cast<arrow::MonthIntervalArray>(array_data);
-  memcpy(kI(k_array), month_array->raw_values(), month_array->length() * sizeof(arrow::MonthIntervalArray::value_type));
+  if( type_overrides.null_mapping.have_month_interval ){
```
Comment (nmcdonnell-kx) [inline src/ArrayReader.cpp:503]:
```suggestion
  if( type_overrides.null_mapping.have_month_interval && month_array->null_count() ){

```

File: src/ArrayReader.cpp (hunk 3 above)
Comment (vgrechin-kx) [reply src/ArrayReader.cpp:249]: Thanks, overlooked!

File: src/KdbOptions.h [hunk 5]
```diff
@@ -709,7 +659,7 @@ inline void KdbOptions::HandleNullMapping<arrow::Type::FIXED_SIZE_BINARY>( const
 template<>
 inline void KdbOptions::HandleNullMapping<arrow::Type::DATE32>( const std::string& key, K value )
 {
-  if( -KI == value->t ){
+  if( value->t == -KI || value->t == -KD ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:662]: I don't think you need to allow -KI here

File: src/KdbOptions.h [hunk 6]
```diff
@@ -721,7 +671,7 @@ inline void KdbOptions::HandleNullMapping<arrow::Type::DATE32>( const std::strin
 template<>
 inline void KdbOptions::HandleNullMapping<arrow::Type::DATE64>( const std::string& key, K value )
 {
-  if( -KJ == value->t ){
+  if( value->t == -KJ || value->t == -KP ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:674]: I don't think you need to allow -KJ here

File: src/KdbOptions.h [hunk 7]
```diff
@@ -733,7 +683,7 @@ inline void KdbOptions::HandleNullMapping<arrow::Type::DATE64>( const std::strin
 template<>
 inline void KdbOptions::HandleNullMapping<arrow::Type::TIMESTAMP>( const std::string& key, K value )
 {
-  if( -KJ == value->t ){
+  if( value->t == -KJ || value->t == -KP ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:686]: I don't think you need to allow -KJ here

File: src/KdbOptions.h [hunk 8]
```diff
@@ -745,7 +695,7 @@ inline void KdbOptions::HandleNullMapping<arrow::Type::TIMESTAMP>( const std::st
 template<>
 inline void KdbOptions::HandleNullMapping<arrow::Type::TIME32>( const std::string& key, K value )
 {
-  if( -KI == value->t ){
+  if( value->t == -KI || value->t == -KT ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:698]: I don't think you need to allow -KI here

File: src/KdbOptions.h [hunk 9]
```diff
@@ -757,7 +707,7 @@ inline void KdbOptions::HandleNullMapping<arrow::Type::TIME32>( const std::strin
 template<>
 inline void KdbOptions::HandleNullMapping<arrow::Type::TIME64>( const std::string& key, K value )
 {
-  if( -KJ == value->t ){
+  if( value->t == -KJ || value->t == -KN ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:710]: I don't think you need to allow -KJ here

File: src/KdbOptions.h [hunk 10]
```diff
@@ -781,7 +731,7 @@ inline void KdbOptions::HandleNullMapping<arrow::Type::DECIMAL>( const std::stri
 template<>
 inline void KdbOptions::HandleNullMapping<arrow::Type::DURATION>( const std::string& key, K value )
 {
-  if( -KJ == value->t ){
+  if( value->t == -KJ || value->t == -KN ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:734]: I don't think you need to allow -KJ here

File: src/KdbOptions.h [hunk 11]
```diff
@@ -793,7 +743,7 @@ inline void KdbOptions::HandleNullMapping<arrow::Type::DURATION>( const std::str
 template<>
 inline void KdbOptions::HandleNullMapping<arrow::Type::INTERVAL_MONTHS>( const std::string& key, K value )
 {
-  if( -KI == value->t ){
+  if( value->t == -KI || value->t == -KM ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:746]: I don't think you need to allow -KI here

File: src/KdbOptions.h [hunk 12]
```diff
@@ -805,7 +755,7 @@ inline void KdbOptions::HandleNullMapping<arrow::Type::INTERVAL_MONTHS>( const s
 template<>
 inline void KdbOptions::HandleNullMapping<arrow::Type::INTERVAL_DAY_TIME>( const std::string& key, K value )
 {
-  if( -KJ == value->t ){
+  if( value->t == -KJ || value->t == -KN ){
```
Comment (nmcdonnell-kx) [inline src/KdbOptions.h:758]: I don't think you need to allow -KJ here

File: src/KdbOptions.h (hunk 5 above)
Comment (vgrechin-kx) [reply src/KdbOptions.h:662]: Yep, looks sub-optimal

**Review by nmcdonnell-kx — APPROVED** [review]

**Pull request merged**